### ISSUE TITLE
[ci] Azure Mariner CI jobs regularly failing: "File not found: 'docker'"

### ISSUE THREAD

**Issue by jameslamb** [microsoft/LightGBM#6316]
## Description

Since switching the `Linux` CI jobs at Azure DevOps to Mariner Linux in #6222, we've seen an increased rate of Azure DevOps jobs failing.

Typically with an error like this in the "initialize job" stage:

```text
##[error]File not found: 'docker'
```

Creating this issue to track those cases.

## Reproducible example

A few recent cases:

* #6308 - `Linux inference` ([build link](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=15766&view=logs&j=a99e48cf-d887-5e89-64ae-c06f3e281b3b&t=ec4d64b2-ec51-4e09-81b1-5fad9a162b2f))
* #6305 - `Linux sdist` ([build link](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=15757&view=logs&j=02a2c3ba-81f8-54e3-0767-5d5adbb0daa9&t=3ff0a57c-0851-47cf-a951-26e1245050f6&s=96ac2280-8cb4-5df5-99de-dd2da759617d))
* #6305 - `Linux_latest bdist` ([build link](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=15757&view=logs&j=b463fb3a-b487-5cfc-fc07-6d216464ba86&t=44e877da-3787-4f96-8a61-e8935660668f&s=96ac2280-8cb4-5df5-99de-dd2da759617d))
* #6305 - `Linux_latest gpu_pip` ([build link](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=15757&view=logs&j=a29125ea-430e-5585-1c1e-7fee1347f663&t=7ae74c60-48d0-473e-bdde-c812d9e79257&s=96ac2280-8cb4-5df5-99de-dd2da759617d))

All had similar logs, like this:

```text
Starting: Initialize job
Agent name: 'lgbub6c540001AS'
Agent machine name: 'lgbub6c540001AS'
Current agent version: '3.234.0'
Agent running as: 'AzDevOps'
Prepare build directory.
Set build variables.
Download all required tasks.
Downloading task: 1ESHostedPoolValidation (1.0.26350327)
Downloading task: CmdLine (2.231.1)
Downloading task: Bash (3.231.5)
Downloading task: PublishBuildArtifacts (1.231.1)
Downloading task: ComponentGovernanceComponentDetection (0.2420208.1)
Checking job knob settings.
   Knob: DockerActionRetries = true Source: $(VSTSAGENT_DOCKER_ACTION_RETRIES) 
   Knob: AgentEnablePipelineArtifactLargeChunkSize = true Source: $(AGENT_ENABLE_PIPELINEARTIFACT_LARGE_CHUNK_SIZE) 
   Knob: ContinueAfterCancelProcessTreeKillAttempt = true Source: $(VSTSAGENT_CONTINUE_AFTER_CANCEL_PROCESSTREEKILL_ATTEMPT) 
   Knob: ProcessHandlerTelemetry = true Source: $(AZP_75787_ENABLE_COLLECT) 
   Knob: IgnoreVSTSTaskLib = true Source: $(AZP_AGENT_IGNORE_VSTSTASKLIB) 
   Knob: FailJobWhenAgentDies = true Source: $(FAIL_JOB_WHEN_AGENT_DIES) 
   Knob: CheckForTaskDeprecation = true Source: $(AZP_AGENT_CHECK_FOR_TASK_DEPRECATION) 
   Knob: MountWorkspace = true Source: $(AZP_AGENT_MOUNT_WORKSPACE) 
Finished checking job knob settings.
##[error]File not found: 'docker'
Finishing: Initialize job
```

## Environment info

N/A

## Additional Comments

Pulling this into its own issue (original conversation started in https://github.com/microsoft/LightGBM/pull/6307#issuecomment-1935610311).

**Comment by jameslamb** [comment]
2 more on #6331:

* #6331 - `Linux_latest gpu_pip` ([build link](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=15798&view=logs&j=a29125ea-430e-5585-1c1e-7fee1347f663))
* #6331 - `Linux_latest gpu_wheel` ([build link](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=15798&view=logs&j=6ee41f30-52b9-5fb2-cd2b-d0121164dc51))

**Comment by jameslamb** [comment]
And one on #6019:

* #6019 - `Linux_latest mpi_wheel` ([build link](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=15799&view=logs&j=fa2f6285-a494-554d-e130-a6421bfeddcc))

**Comment by shiyu1994** [comment]
I will investigate this. Thanks for creating the issue.

**Comment by jameslamb** [comment]
I'm not going to post as many of these since I think it's clear at this point that there's an issue.

But want to share this from a run on #6341 about 16 hours ago... 8 of 18 jobs **on the same run** failed at initialization with the error reported in the original post in this issue.

<img width="630" alt="image" src="https://github.com/microsoft/LightGBM/assets/7608904/8e7bccc5-c08b-445c-8691-a23114458635">

<img width="1123" alt="image" src="https://github.com/microsoft/LightGBM/assets/7608904/d706173e-f190-4cad-a49c-52d70dd9594e">

([build link](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=15856&view=results))

**Comment by jameslamb** [comment]
In addition to these frequent failures, over the last few days I've observed what looks like significantly reduced capacity.

For example, the Linux tasks in this build have all been stuck in `Queued` for 11 hours: https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=15983&view=results. Despite there being 0 other commits currently building on Azure DevOps in the LightGBM project.

@shiyu1994 could you look into that? Is LightGBM competing with other pipelines for capacity?

**Comment by jameslamb** [comment]
I just restarted the jobs on #6364 and saw all the Linux jobs on Azure DevOps get queued, with messages like this

```text
The agent request is not running because all potential agents are running other requests. Current position in queue: 131
```

Some jobs have been stuck in "queued", waiting to be picked up, for **more than 4 days**.

<img width="1426" alt="Screenshot 2024-03-28 at 9 48 57 PM" src="https://github.com/microsoft/LightGBM/assets/7608904/27e8b6e7-6a58-43fe-8b8d-fe5a14f99217">

@shiyu1994 can you please look into this? Is LightGBM competing with other projects, or is Azure's capacity for these types of VMs just very limited?

We really rely heavily on these Linux jobs on Azure DevOps and this disruption is blocking development on the project.

**Comment by jameslamb** [comment]
I've added the `blocking` label. It's now been 5+ days since we had a single successful run of CI.

#6394 was opened about 12 hours ago, for example, and all of its Linux CI jobs are stuck in `Queued`: https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=16010&view=results

These jobs just not being rung at all is different from the other issue reported in the first post here (`file not found: docker`), but I'm including this here so all discussion about instability with these new runners can be kept in one place.

**Comment by jameslamb** [comment]
Thanks! Whenever it's fixed, I can take care of rebuilding + merging all the already-approved PRs.

**Comment by jameslamb** [comment]
Over the last 2 weeks at least, I have not seen the "File not found: 'docker'" issue a single time! 🎉 

It's becoming much more common that all of CI across all providers passes with 0 manual re-runs, as just happened on `master` after I merged #6398.

I wonder if some combination of #6407, #6416, and other fixes made by Azure have all contributed to stabilizing this. Either way, I think we can close this issue for now and re-open it if the problems come up again.

Thanks for all your help @shiyu1994 !!!
